### DISCUSSION TITLE
Long running process and systemd alternative for Codespaces

### DISCUSSION THREAD

**Discussion by lucernae** [community/community#6242] [Codespaces]
Hi, I'm part of the codespace beta tester. So far I really like this experience.

I have some question regarding a specific use case. Can you tell me if this is going to be supported or is not possible

My use case is that I want to spinup codespace so that it immediately setup a kubernetes cluster (k3s in this context), so that I can test my work in this codespace (can be helm charts, or simulated end-to-end tests of a microservices setup).
So far I can instantiate k3s. It easily simplify my work because I can just create/delete small Kubernetes cluster on-demand and cleanly organize it (separated by branch). I believe this approach will provide value or even a new ecosystem in GitHub for cloud native projects.

My question regarding this:
- Is it possible to disable the suspend logic? My tests can run in a long time, possibly more than 30 minutes if it includes full build within the codespace. I also want to start the codespace and then use it as some kind of mini-cloud-worker that can be triggered remotely from a CI pipeline, so I need the the k8s endpoint to be up even if I close my browser tab.
- Is it possible to have SSL passthru in the https port of the autogenerated port forward from Codespace? Does the https support websocket?
- Are there any API that can control this codespace behaviour or environment. Or can query the state of the codespace (if it is suspended, start the codespace remotely via API or something). I also want to query the generated DNS name that it provide for the port forward
- I know that systemd is disabled in Codespace. Are there any alternatives provided for Codespace to run user-specified services when the codespace start? Technically, how do you start Docker daemon if you don't have systemd? I would like to use the same approach.

**Comment by Chuxel** [comment]
For the suspend, if terminal output is going on in the integrated terminal, you shouldn't see the codespace getting suspended. Unfortunately, you can't override it right now.


If you're using VS Code locally with the Codespaces extension, TCP port are mapped directly to localhost. As a result, SSL traffic flows as-is. You can use your hosts file to mimic another domain if needed. This is not supported when using the web URI currently.

In terms of startup, there are a few tips for doing this here: https://code.visualstudio.com/remote/advancedcontainers/start-processes If you'd prefer to have a manager like `systemd` instead, [Supervisor](http://supervisord.org/) is an alternative that focuses on process startup without all of the rest of the complexity that makes systemd slow starting and is really appropriate for physical hardware or multi-user systems anyway.  You'd just have your entrypoint start `supervisord` in this case.

API support is coming - so stay tuned!

**Reply by lucernae** [reply]
> In terms of startup, there are a few tips for doing this here: https://code.visualstudio.com/remote/advancedcontainers/start-processes If you'd prefer to have a manager like systemd instead, Supervisor is an alternative that focuses on process startup without all of the rest of the complexity that makes systemd slow starting and is really appropriate for physical hardware or multi-user systems anyway. You'd just have your entrypoint start supervisord in this case.

This one seems to cover my use case.

> If you're using VS Code locally with the Codespaces extension, TCP port are mapped directly to localhost. As a result, SSL traffic flows as-is. You can use your hosts file to mimic another domain if needed. This is not supported when using the web URI currently.

I'm aware that tcp are forwarded as is in VS Code local. But it would be super simple for developer experience if we can use Codespace's generated DNS name/url and make it accept websockets. I think, only the websocket support is missing. Can you clarifiy this?

> API support is coming - so stay tuned!

Awesome, looking forward to this.

**Reply by Chuxel** [reply]
@lucernae If you are talking about on the web, I'm not aware of any issues around web sockets. However, if you are using a different port, that will end up at a different URL. Can you be more specific?
